File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def solution(maps):
    q = deque([(0, 0, 1)])
    n = len(maps)
    m = len(maps[0])
    dx = [1, -1, 0, 0]
    dy = [0, 0, 1, -1]
    visited = [[False] * m for _ in range(n)]
    visited[0][0] = True
    
    while q:
        x, y, total = q.popleft()
        
        if x == n-1 and y == m-1:
            return total
        
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
            
            if 0 <= nx and nx < n and 0 <= ny and ny < m and not visited[nx][ny] and maps[nx][ny] == 1:
                q.append((nx, ny, total+1))
                visited[nx][ny] = True
        
    return -1

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;

class Solution {
    public int solution(int[][] maps) {
        int n = maps.length;
        int m = maps[0].length;
        boolean[][] visited = new boolean[n][m];
        
        Queue<int[]> q = new LinkedList<>();
        q.add(new int[] {0, 0, 1});
        visited[0][0] = true;
        
        int[] dx = {-1, 1, 0, 0};
        int[] dy = {0, 0, 1, -1};
        
        while (!q.isEmpty()){
            int[] curr = q.poll();
            int x = curr[0];
            int y = curr[1];
            int dist = curr[2];
            
            if (x == n-1 && y == m-1) return dist;
            
            for (int i = 0; i< 4; i++){
                int nx = x + dx[i];
                int ny = y + dy[i];
                
                if (nx >= 0 && nx < n && ny >= 0 && ny < m && maps[nx][ny] == 1 && !visited[nx][ny]){
                    q.add(new int[] {nx, ny, dist+1});
                    visited[nx][ny] = true;
                }
            }
        }
        return -1;
    }
}